include/hip/hcc_detail/math_functions.h

```diff
@@ -20,237 +20,1125 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+inline
+float rhypotf(float x, float y) { return __ocml_rhypot_f32(x, y); }
+__device__
+inline
+float rintf(float x) { return __ocml_rint_f32(x); }
+__device__
+inline
+float rnorm3df(float x, float y, float z)
+{
+    return __ocml_rlen3_f32(x, y, z);
+}
+
+__device__
+inline
+float rnorm4df(float x, float y, float z, float w)
+{
+    return __ocml_rlen4_f32(x, y, z, w);
+}
+__device__
+inline
+float rnormf(int dim, const float* a)
+{
+    float r = 0;
+    while (dim--) { r += a[0] * a[0]; ++a; }
+
+    return __ocml_rsqrt_f32(r);
+}
+__device__
+inline
+float roundf(float x) { return __ocml_round_f32(x); }
+__device__
+inline
+float rsqrtf(float x) { return __ocml_rsqrt_f32(x); }
+__device__
+inline
+float scalblnf(float x, long int n)
+{
+    return (n < INT_MAX) ? __ocml_scalbn_f32(x, n) : __ocml_scalb_f32(x, n);
+}
+__device__
+inline
+float scalbnf(float x, int n) { return __ocml_scalbn_f32(x, n); }
+__device__
+inline
+int signbit(float x) { return __ocml_signbit_f32(x); }
+__device__
+inline
+void sincosf(float x, float* sptr, float* cptr)
+{
+    float tmp;
+
+    *sptr =
+        __ocml_sincos_f32(x, (__attribute__((address_space(5))) float*) &tmp);
+    *cptr = tmp;
+}
+__device__
+inline
+void sincospif(float x, float* sptr, float* cptr)
+{
+    float tmp;
+
+    *sptr =
+        __ocml_sincospi_f32(x, (__attribute__((address_space(5))) float*) &tmp);
+    *cptr = tmp;
+}
+__device__
+inline
+float sinf(float x) { return __ocml_sin_f32(x); }
+__device__
+inline
+float sinhf(float x) { return __ocml_sinh_f32(x); }
+__device__
+inline
+float sinpif(float x) { return __ocml_sinpi_f32(x); }
+__device__
+inline
+float sqrtf(float x) { return __ocml_sqrt_f32(x); }
+__device__
+inline
+float tanf(float x) { return __ocml_tan_f32(x); }
+__device__
+inline
+float tanhf(float x) { return __ocml_tanh_f32(x); }
+__device__
+inline
+float tgammaf(float x) { return __ocml_tgamma_f32(x); }
+__device__
+inline
+float truncf(float x) { return __ocml_trunc_f32(x); }
+__device__
+inline
+float y0f(float x) { return __ocml_y0_f32(x); }
+__device__
+inline
+float y1f(float x) { return __ocml_y1_f32(x); }
+__device__
+inline
+float ynf(int n, float x)
+{   // TODO: we could use Ahmes multiplication and the Miller & Brown algorithm
+    //       for linear recurrences to get O(log n) steps, but it's unclear if
+    //       it'd be beneficial in this case.
+    if (n == 0) return y0f(x);
+    if (n == 1) return y1f(x);
+
+    float x0 = y0f(x);
+    float x1 = y1f(x);
+    for (int i = 1; i < n; ++i) {
+        float x2 = (2 * i) / x * x1 - x0;
+        x0 = x1;
+        x1 = x2;
+    }
+
+    return x1;
+}
+
+// BEGIN INTRINSICS
+__device__
+inline
+float __cosf(float x) { return __llvm_amdgcn_cos_f32(x); }
+__device__
+inline
+float __exp10f(float x) { return __ocml_exp10_f32(x); }
+__device__
+inline
+float __expf(float x) { return __ocml_exp_f32(x); }
+__device__
+inline
+float __fadd_rd(float x, float y) { return __llvm_add_rte_f32(x, y); }
```

Comment:
These special rounding operations are in OCML and should be used from there.  BTW, they do not work currently.